### DISCUSSION TITLE
Change outgoing IP address of connections from the server

### DISCUSSION THREAD

**Discussion by Shayan-To** [shadowsocks/shadowsocks-rust#1070] [Q&A] [answered]
I have a server with two IP addresses, and I want to set up two ShadowSocks servers, each working on one of the IP addresses. I've set their `server` to the two addresses, but it only changes the address the servers binds to. When I set my proxy and check the IP address that is seen by the remote servers (for example by `https_proxy='socks5h://localhost:1080' curl https://ifconfig.co`), it is the same for both of the ShadowSocks servers.

What should I set in the config to fix this?

**Comment by Shayan-To** [comment]
@zonyitoo Sorry to bother you, but could you please take a look at my problem?

**Comment by zonyitoo** [comment]
So you have already set the `server` to the IP address that your server instance binds to, then the socket was already `bind()`ing and `listen()`ing to the specific IP address. What do you mean "it is the same for both of the ShadowSocks servers"? What's the problem you have encountered actually?

**Reply by Shayan-To** [reply]
Let me try to be more clear.

I have a server with two IPs, IP1 and IP2. I start an `ssserver` with `server` set to IP2. On my local machine, I start an `sslocal` that connects to IP2, and I set `https_proxy` to go through my `sslocal`. Then I use `curl` to query my IP address from some remote service (like `https://ifconfig.co`). So I call `https_proxy='socks5h://localhost:1080' curl https://ifconfig.co` and the remote server returns **IP1** to me. I want to change this, so that the remote server sees me as IP2, not as IP1.

In other words, `ssserver` is binding to and listening on IP2, but when I request a connection from my `sslocal`, then `ssserver` uses **IP1** to connect to the requested destination. What I need is to instruct `ssserver` to use IP2 to connect to the requested destinations, so the remote addresses I connect to, see me as IP2.

Could I put what I mean to words clearly?

**Reply by zonyitoo** [reply]
Add `local_address` in configuration, or `--outbound-bind-interface` on command line arguments to let `ssserver`'s outbound socket binds to the requested interface.

**Reply by Shayan-To** [reply]
I've already tried adding `local_address`, but nothing changes. I just tried it again with the latest version, with both gnu and musl builds, and they both log this:

    [shadowsocks_service::server::tcprelay] established tcp tunnel ....:... <-> ifconfig.co:443 with ConnectOpts { fwmark: None, bind_local_addr: None, bind_interface: None, tcp: TcpSocketOpts { send_buffer_size: None, recv_buffer_size: None, nodelay: false, fastopen: true, keepalive: Some(15s) } }

And the remote server sees me as IP1. (I've set both `server` and `local_adderss` to IP2.)

I'll try to find out how to use `--outbound-bind-interface` now and report the results.

**Reply by Shayan-To** [reply]
It seems like I have a single network interface `venet0` that controls the both addresses. `ifconfig` is showing me `venet0:0` for my IP1 and `venet0:1` for my IP2.

I tried `--outbound-bind-interface venet0:1`, and this error was logged:

    [shadowsocks::net::sys::unix::linux] set SO_BINDTODEVICE error: No such device (os error 19)

I tried `--outbound-bind-interface venet0`, everything was the same as before, except the log message that contained `bind_interface: Some("venet0")`.

**Reply by Shayan-To** [reply]
I found `--outbound-bind-addr` in `ssserver --help`, and it worked!
Is there a way to specify this in the config file?

**Reply by zonyitoo** [reply]
You can use `outbound_bind_addr` with the latest commit.

**Reply by Shayan-To** [reply]
Thanks a lot for the quick update.

A little feedback: In `README`, in the code block right above the one you added the `outbound_...` configuration examples, there is this comment and config:

```
    // Outbound socket binds to this IP address
    // For choosing different network interface on the same machine
    "local_address": "xxx.xxx.xxx.xxx",
```

Is this info incorrect or am I reading it wrong?

**Comment by dev4u** [comment]
我觉得你的问题是想搞负载均衡？robin方式的负载？
据我所知，rust的多服务端工作模式，是类似多机互备的机制：master不宕机，是不会切换到slave上的。
有两个建议可以试试：
- 用haproxy之类的软件来做前端
- 修改rust主备切换逻辑

不管用什么方式，你可能会遇到被浏览网站判定为机器人，而引导去做拼图游戏……甚至拒绝服务。

**Comment by Shayan-To** [comment, marked as answer]
To summarize:
Use `--outbound-bind-addr` command line argument.
Or with the latest commits, use `outbound_bind_addr` in the config file.
